Fix qgsVariantGreaterThan may return false result when values are equal

nyalldawson opened the pull request:
Fixes #61759

nyalldawson committed:
Fix qgsVariantGreaterThan may return false result when values are equal

Fixes #61759

rouault commented on src/core/qgis.cpp:
```diff
@@ -126,52 +126,115 @@ void qgsFree( void *ptr )
+    {
+      const double lhsDouble = lhs.toDouble();
+      const double rhsDouble = rhs.toDouble();
+      return lhsDouble < rhsDouble ? -1 : ( lhsDouble == rhsDouble ? 0 : 1 );
```
what about NaN comparisons ? (also true for Float).

I'd suggest considering NaN < any non-NaN value.

So 
```suggestion
      if ( std::isnan(lhsDouble) )
      {
           return std::isnan(rhsDouble) ? 0 : -1;
      }
      else if ( std::isnan(rhsDouble) )
      {
          return 1;
      }
      return lhsDouble < rhsDouble ? -1 : ( lhsDouble == rhsDouble ? 0 : 1 );
```

nyalldawson replied:
Ok, added

nyalldawson committed:
Handle nan float/double in comparison

nyalldawson committed:
Fix test

The test was not deterministic, as it relies on sorting a list
containing two identical values and which may be validly sorted
in either order
